### PR TITLE
Make model tests name clear

### PR DESCRIPTION
**Pull request by chilo-ms**
This PR makes model tests show meaningful and clear test name instead of sequential number so user can quickly know which tests are being run. 

original test name: 

    ...
    1: [ RUN      ] ModelTests/ModelTest.Run/6990
    1: [       OK ] ModelTests/ModelTest.Run/6990 (0 ms)
    1: [ RUN      ] ModelTests/ModelTest.Run/6991
    1: [       OK ] ModelTests/ModelTest.Run/6991 (0 ms)
    1: [ RUN      ] ModelTests/ModelTest.Run/6992

new test name:

    ...
    1: [ RUN      ] ModelTests/ModelTest.Run/tensorrt__data_onnx_onnx123_node_test_hardmax_default_axis
    1: [       OK ] ModelTests/ModelTest.Run/tensorrt__data_onnx_onnx123_node_test_hardmax_default_axis (0 ms)
    1: [ RUN      ] ModelTests/ModelTest.Run/tensorrt__data_onnx_onnx123_node_test_averagepool_1d_default
    1: [       OK ] ModelTests/ModelTest.Run/tensorrt__data_onnx_onnx123_node_test_averagepool_1d_default (0 ms)

### CODE REVIEW

**Comment by jywu-msft** [comment]
Thanks! this is very useful

File: onnxruntime/test/providers/cpu/model_tests.cc [hunk 1]
```diff
@@ -990,7 +990,43 @@ ::std::vector<::std::basic_string<ORTCHAR_T>> GetParameterStrings() {
   return v;
 }
 
+auto GenerateCustomTestName  = [](const ::testing::TestParamInfo<ModelTest::ParamType>& info) {
```
Comment (jywu-msft) [inline onnxruntime/test/providers/cpu/model_tests.cc:993]: thinking there could be a better name? 
these are not really custom tests. 
something like ExpandModelName or GenerateModelName ?

Comment (chilo-ms) [reply onnxruntime/test/providers/cpu/model_tests.cc]: Changed it to ExpandModelName

**Comment by chilo-ms** [comment]
From the define of INSTANTIATE_TEST_SUITE_P we know that the last argument must be a function that returns a type of std::string. For Windows, we uses std::wstring, so I don't think we can use this argument on Windows for now.

   

     // The last argument to INSTANTIATE_TEST_SUITE_P allows the user to specify
    // generator and an optional function or functor that generates custom test name
    // suffixes based on the test parameters. Such a function or functor should
    // accept one argument of type testing::TestParamInfo<class ParamType>, and
    // return std::string.
    ...
    #define INSTANTIATE_TEST_SUITE_P(prefix, test_suite_name, ...)                \
      static ::testing::internal::ParamGenerator<test_suite_name::ParamType>      \
          gtest_##prefix##test_suite_name##_EvalGenerator_() {                    \
        return GTEST_EXPAND_(GTEST_GET_FIRST_(__VA_ARGS__, DUMMY_PARAM_));        \
      }                                                                           \
      **static ::std::string** gtest_##prefix##test_suite_name##_EvalGenerateName_(   \
          const ::testing::TestParamInfo<test_suite_name::ParamType>& info) {     \
        if (::testing::internal::AlwaysFalse()) {                                 \
          ::testing::internal::TestNotEmpty(GTEST_EXPAND_(GTEST_GET_SECOND_(      \
              __VA_ARGS__,                                                        \
              ::testing::internal::DefaultParamName<test_suite_name::ParamType>,  \
              DUMMY_PARAM_)));                                                    \
          auto t = std::make_tuple(__VA_ARGS__);                                  \
          static_assert(std::tuple_size<decltype(t)>::value <= 2,                 \
                        "Too Many Args!");                                        \
        } 
    ...

**Comment by jywu-msft** [comment]
> From the define of INSTANTIATE_TEST_SUITE_P we know that the last argument must be a function that returns a type of std::string. For Windows, we use std::wstring, so I don't think we can use this argument on Windows for now.
> 
> ```
>  // The last argument to INSTANTIATE_TEST_SUITE_P allows the user to specify
> // generator and an optional function or functor that generates custom test name
> // suffixes based on the test parameters. Such a function or functor should
> // accept one argument of type testing::TestParamInfo<class ParamType>, and
> // return std::string.
> ...
> #define INSTANTIATE_TEST_SUITE_P(prefix, test_suite_name, ...)                \
>   static ::testing::internal::ParamGenerator<test_suite_name::ParamType>      \
>       gtest_##prefix##test_suite_name##_EvalGenerator_() {                    \
>     return GTEST_EXPAND_(GTEST_GET_FIRST_(__VA_ARGS__, DUMMY_PARAM_));        \
>   }                                                                           \
>   static ::std::string gtest_##prefix##test_suite_name##_EvalGenerateName_(   \
>       const ::testing::TestParamInfo<test_suite_name::ParamType>& info) {     \
>     if (::testing::internal::AlwaysFalse()) {                                 \
>       ::testing::internal::TestNotEmpty(GTEST_EXPAND_(GTEST_GET_SECOND_(      \
>           __VA_ARGS__,                                                        \
>           ::testing::internal::DefaultParamName<test_suite_name::ParamType>,  \
>           DUMMY_PARAM_)));                                                    \
>       auto t = std::make_tuple(__VA_ARGS__);                                  \
>       static_assert(std::tuple_size<decltype(t)>::value <= 2,                 \
>                     "Too Many Args!");                                        \
>     } 
> ...                              
> ```

is it possible to convert wstring to std::string to enable on windows?

**Comment by chilo-ms** [comment]
> > From the define of INSTANTIATE_TEST_SUITE_P we know that the last argument must be a function that returns a type of std::string. For Windows, we use std::wstring, so I don't think we can use this argument on Windows for now.
> > ```
> >  // The last argument to INSTANTIATE_TEST_SUITE_P allows the user to specify
> > // generator and an optional function or functor that generates custom test name
> > // suffixes based on the test parameters. Such a function or functor should
> > // accept one argument of type testing::TestParamInfo<class ParamType>, and
> > // return std::string.
> > ...
> > #define INSTANTIATE_TEST_SUITE_P(prefix, test_suite_name, ...)                \
> >   static ::testing::internal::ParamGenerator<test_suite_name::ParamType>      \
> >       gtest_##prefix##test_suite_name##_EvalGenerator_() {                    \
> >     return GTEST_EXPAND_(GTEST_GET_FIRST_(__VA_ARGS__, DUMMY_PARAM_));        \
> >   }                                                                           \
> >   static ::std::string gtest_##prefix##test_suite_name##_EvalGenerateName_(   \
> >       const ::testing::TestParamInfo<test_suite_name::ParamType>& info) {     \
> >     if (::testing::internal::AlwaysFalse()) {                                 \
> >       ::testing::internal::TestNotEmpty(GTEST_EXPAND_(GTEST_GET_SECOND_(      \
> >           __VA_ARGS__,                                                        \
> >           ::testing::internal::DefaultParamName<test_suite_name::ParamType>,  \
> >           DUMMY_PARAM_)));                                                    \
> >       auto t = std::make_tuple(__VA_ARGS__);                                  \
> >       static_assert(std::tuple_size<decltype(t)>::value <= 2,                 \
> >                     "Too Many Args!");                                        \
> >     } 
> > ...                              
> > ```
> 
> is it possible to convert wstring to std::string to enable on windows?
Thanks for the suggestion. It's possible and I added the conversion to the code.

**Review by jywu-msft — APPROVED** [review]

**Comment by jywu-msft** [comment]
@snnn can you take a look at this PR?

File: onnxruntime/test/providers/cpu/model_tests.cc [hunk 2]
```diff
@@ -990,7 +993,43 @@ ::std::vector<::std::basic_string<ORTCHAR_T>> GetParameterStrings() {
...
+  // Note: test name only accepts '_' and alphanumeric
+  // remove '.' and '-'
+  char chars[] = ".-";
+  for (unsigned int i = 0; i < strlen(chars); ++i) {
+    name.erase(std::remove(name.begin(), name.end(), chars[i]), name.end());
+  }
+#ifdef _WIN32
+  // Note: The return value of INSTANTIATE_TEST_SUITE_P accpets std::basic_string<char...>.
+  // Need conversion of wchar_t to char.
+  return std::wstring_convert<std::codecvt_utf8<wchar_t>>().to_bytes(name);
```
Comment (snnn) [inline onnxruntime/test/providers/cpu/model_tests.cc:1024]: Just a tiny comment: internally ORT should only use UTF-8, but when you pass a string outside, for example, feed it to std::cout or a third-party library, you should use a local code page(like ISO-8859-* , GBK), not UTF-8.
Right now our model names do not contain non-ASCII string. In the future it won't too. So, it's fine to  use UTF-8 here. We don't need to make it support non-ASCII strings.

**Review by snnn — APPROVED** [review]

**Comment by snnn** [comment]
The reason I didn't make it meaningful was because I was not sure if model names could contain sensitive information.

**Comment by jywu-msft** [comment]
> The reason I didn't make it meaningful was because I was not sure if model names could contain sensitive information.

2 questions:
1) do we run any non-public tests in our CI's. i thought these were just onnx node tests and models from onnx model zoo which are all public?
2) if there are non-public tests, can we rename any sensitive test filenames? we need a convenient way for developers to reproduce issues - so knowing the real filename of the test is very useful.

**Comment by snnn** [comment]
> > The reason I didn't make it meaningful was because I was not sure if model names could contain sensitive information.
> 
> 2 questions:
> 
> 1. do we run any non-public tests in our CI's. i thought these were just onnx node tests and models from onnx model zoo which are all public?
> 2. if there are non-public tests, can we rename any sensitive test filenames? we need a convenient way for developers to reproduce issues - so knowing the real filename of the test is very useful.

1. Yes
2. Yes

**Comment by chilo-ms** [comment]
Checked with AML team regarding revealing some non-public model names, for examples:
winmlperf_coreml_FNS-Candy_prerelease.onnx
winmlperf_coreml_FNS-Candy_prerelease_fp16.onnx
winmlperf_coreml_Inceptionv3_prerelease.onnx

winmlperf_coreml_LinearRegression_NYCTaxi_prerelease_fp16.onnx

winmlperf_coreml_Resnet50_prerelease.onnx


They are fine with it.

**Pull request merged**